File: akka-cluster-sharding-typed/src/test/java/jdocs/delivery/WorkPullingDocExample.java
/*
 * Copyright (C) 2020-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.delivery;

// #imports
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.delivery.ConsumerController;
import akka.actor.typed.delivery.DurableProducerQueue;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;

import java.time.Duration;
import java.util.Optional;
import java.util.UUID;

// #imports

// #producer
import akka.actor.typed.delivery.WorkPullingProducerController;
import akka.Done;

// #producer

// #durable-queue
import akka.persistence.typed.PersistenceId;
import akka.persistence.typed.delivery.EventSourcedProducerQueue;

// #durable-queue

import akka.actor.typed.javadsl.StashBuffer;
import akka.actor.typed.receptionist.ServiceKey;

interface WorkPullingDocExample {

  // #consumer
  public class ImageConverter {
    interface Command {}

    public static class ConversionJob {
      public final UUID resultId;
      public final String fromFormat;
      public final String toFormat;
      public final byte[] image;

      public ConversionJob(UUID resultId, String fromFormat, String toFormat, byte[] image) {
        this.resultId = resultId;
        this.fromFormat = fromFormat;
        this.toFormat = toFormat;
        this.image = image;
      }
    }

    private static class WrappedDelivery implements Command {
      final ConsumerController.Delivery<ConversionJob> delivery;

      private WrappedDelivery(ConsumerController.Delivery<ConversionJob> delivery) {
        this.delivery = delivery;
      }
    }

    public static ServiceKey<ConsumerController.Command<ConversionJob>> serviceKey =
        ServiceKey.create(ConsumerController.serviceKeyClass(), "ImageConverter");

    public static Behavior<Command> create() {
      return Behaviors.setup(
          context -> {
            ActorRef<ConsumerController.Delivery<ConversionJob>> deliveryAdapter =
                context.messageAdapter(ConsumerController.deliveryClass(), WrappedDelivery::new);
            ActorRef<ConsumerController.Command<ConversionJob>> consumerController =
                context.spawn(ConsumerController.create(serviceKey), "consumerController");
            consumerController.tell(new ConsumerController.Start<>(deliveryAdapter));

            return Behaviors.receive(Command.class)
                .onMessage(WrappedDelivery.class, ImageConverter::onDelivery)
                .build();
          });
    }

    private static Behavior<Command> onDelivery(WrappedDelivery w) {
      byte[] image = w.delivery.message().image;
      String fromFormat = w.delivery.message().fromFormat;
      String toFormat = w.delivery.message().toFormat;
      // convert image...
      // store result with resultId key for later retrieval

      // and when completed confirm
      w.delivery.confirmTo().tell(ConsumerController.confirmed());

      return Behaviors.same();
    }
  }
  // #consumer

  // #producer
  public class ImageWorkManager {

    interface Command {}

    public static class Convert implements Command {
      public final String fromFormat;
      public final String toFormat;
      public final byte[] image;

      public Convert(String fromFormat, String toFormat, byte[] image) {
        this.fromFormat = fromFormat;
        this.toFormat = toFormat;
        this.image = image;
      }
    }

    public static class GetResult implements Command {
      public final UUID resultId;
      public final ActorRef<Optional<byte[]>> replyTo;

      public GetResult(UUID resultId, ActorRef<Optional<byte[]>> replyTo) {
        this.resultId = resultId;
        this.replyTo = replyTo;
      }
    }

    private static class WrappedRequestNext implements Command {
      final WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next;

      private WrappedRequestNext(
          WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next) {
        this.next = next;
      }
    }

    // #producer
    // #ask
    public static class ConvertRequest implements Command {
      public final String fromFormat;
      public final String toFormat;
      public final byte[] image;
      public final ActorRef<ConvertResponse> replyTo;

      public ConvertRequest(
          String fromFormat, String toFormat, byte[] image, ActorRef<ConvertResponse> replyTo) {
        this.fromFormat = fromFormat;
        this.toFormat = toFormat;
        this.image = image;
        this.replyTo = replyTo;
      }
    }

    interface ConvertResponse {}

    public static class ConvertAccepted implements ConvertResponse {
      public final UUID resultId;

      public ConvertAccepted(UUID resultId) {
        this.resultId = resultId;
      }
    }

    enum ConvertRejected implements ConvertResponse {
      INSTANCE
    }

    public static class ConvertTimedOut implements ConvertResponse {
      public final UUID resultId;

      public ConvertTimedOut(UUID resultId) {
        this.resultId = resultId;
      }
    }

    private static class AskReply implements Command {
      final UUID resultId;
      final ActorRef<ConvertResponse> originalReplyTo;
      final boolean timeout;

      private AskReply(UUID resultId, ActorRef<ConvertResponse> originalReplyTo, boolean timeout) {
        this.resultId = resultId;
        this.originalReplyTo = originalReplyTo;
        this.timeout = timeout;
      }
    }

    // #ask
    // #producer

    private final ActorContext<Command> context;
    private final StashBuffer<Command> stashBuffer;

    private ImageWorkManager(ActorContext<Command> context, StashBuffer<Command> stashBuffer) {
      this.context = context;
      this.stashBuffer = stashBuffer;
    }

    public static Behavior<Command> create() {
      return Behaviors.setup(
          context -> {
            ActorRef<WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob>>
                requestNextAdapter =
                    context.messageAdapter(
                        WorkPullingProducerController.requestNextClass(), WrappedRequestNext::new);
            ActorRef<WorkPullingProducerController.Command<ImageConverter.ConversionJob>>
                producerController =
                    context.spawn(
                        WorkPullingProducerController.create(
                            ImageConverter.ConversionJob.class,
                            "workManager",
                            ImageConverter.serviceKey,
                            Optional.empty()),
                        "producerController");
            // #producer
            // #durable-queue
            Behavior<DurableProducerQueue.Command<ImageConverter.ConversionJob>> durableQueue =
                EventSourcedProducerQueue.create(PersistenceId.ofUniqueId("ImageWorkManager"));
            ActorRef<WorkPullingProducerController.Command<ImageConverter.ConversionJob>>
                durableProducerController =
                    context.spawn(
                        WorkPullingProducerController.create(
                            ImageConverter.ConversionJob.class,
                            "workManager",
                            ImageConverter.serviceKey,
                            Optional.of(durableQueue)),
                        "producerController");
            // #durable-queue
            // #producer
            producerController.tell(new WorkPullingProducerController.Start<>(requestNextAdapter));

            return Behaviors.withStash(
                1000, stashBuffer -> new ImageWorkManager(context, stashBuffer).waitForNext());
          });
    }

    private Behavior<Command> waitForNext() {
      return Behaviors.receive(Command.class)
          .onMessage(WrappedRequestNext.class, this::onWrappedRequestNext)
          .onMessage(Convert.class, this::onConvertWait)
          .onMessage(GetResult.class, this::onGetResult)
          .build();
    }

    private Behavior<Command> onWrappedRequestNext(WrappedRequestNext w) {
      return stashBuffer.unstashAll(active(w.next));
    }

    private Behavior<Command> onConvertWait(Convert convert) {
      if (stashBuffer.isFull()) {
        context.getLog().warn("Too many Convert requests.");
        return Behaviors.same();
      } else {
        stashBuffer.stash(convert);
        return Behaviors.same();
      }
    }

    private Behavior<Command> onGetResult(GetResult get) {
      // TODO retrieve the stored result and reply
      return Behaviors.same();
    }

    private Behavior<Command> active(
        WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next) {
      return Behaviors.receive(Command.class)
          .onMessage(Convert.class, c -> onConvert(c, next))
          .onMessage(GetResult.class, this::onGetResult)
          .onMessage(WrappedRequestNext.class, this::onUnexpectedWrappedRequestNext)
          .build();
    }

    private Behavior<Command> onUnexpectedWrappedRequestNext(WrappedRequestNext w) {
      throw new IllegalStateException("Unexpected RequestNext");
    }

    private Behavior<Command> onConvert(
        Convert convert,
        WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next) {
      UUID resultId = UUID.randomUUID();
      next.sendNextTo()
          .tell(
              new ImageConverter.ConversionJob(
                  resultId, convert.fromFormat, convert.toFormat, convert.image));
      return waitForNext();
    }
    // #producer

    Object askScope =
        new Object() {
          // #ask
          private Behavior<Command> waitForNext() {
            return Behaviors.receive(Command.class)
                .onMessage(WrappedRequestNext.class, this::onWrappedRequestNext)
                .onMessage(ConvertRequest.class, this::onConvertRequestWait)
                .onMessage(AskReply.class, this::onAskReply)
                .onMessage(GetResult.class, this::onGetResult)
                .build();
          }

          private Behavior<Command> onConvertRequestWait(ConvertRequest convert) {
            if (stashBuffer.isFull()) {
              convert.replyTo.tell(ConvertRejected.INSTANCE);
              return Behaviors.same();
            } else {
              stashBuffer.stash(convert);
              return Behaviors.same();
            }
          }

          private Behavior<Command> onAskReply(AskReply reply) {
            if (reply.timeout) reply.originalReplyTo.tell(new ConvertTimedOut(reply.resultId));
            else reply.originalReplyTo.tell(new ConvertAccepted(reply.resultId));
            return Behaviors.same();
          }

          private Behavior<Command> onWrappedRequestNext(WrappedRequestNext w) {
            return stashBuffer.unstashAll(active(w.next));
          }

          private Behavior<Command> onGetResult(GetResult get) {
            // TODO retrieve the stored result and reply
            return Behaviors.same();
          }

          private Behavior<Command> active(
              WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next) {
            return Behaviors.receive(Command.class)
                .onMessage(ConvertRequest.class, c -> onConvertRequest(c, next))
                .onMessage(AskReply.class, this::onAskReply)
                .onMessage(GetResult.class, this::onGetResult)
                .onMessage(WrappedRequestNext.class, this::onUnexpectedWrappedRequestNext)
                .build();
          }

          private Behavior<Command> onConvertRequest(
              ConvertRequest convert,
              WorkPullingProducerController.RequestNext<ImageConverter.ConversionJob> next) {
            UUID resultId = UUID.randomUUID();

            context.ask(
                Done.class,
                next.askNextTo(),
                Duration.ofSeconds(5),
                askReplyTo ->
                    new WorkPullingProducerController.MessageWithConfirmation<>(
                        new ImageConverter.ConversionJob(
                            resultId, convert.fromFormat, convert.toFormat, convert.image),
                        askReplyTo),
                (done, exc) -> {
                  if (exc == null) return new AskReply(resultId, convert.replyTo, false);
                  else return new AskReply(resultId, convert.replyTo, true);
                });

            return waitForNext();
          }

          private Behavior<Command> onUnexpectedWrappedRequestNext(WrappedRequestNext w) {
            throw new IllegalStateException("Unexpected RequestNext");
          }

          // #ask
        };
    // #producer
  }
  // #producer

}

File: akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala
/*
 * Copyright (C) 2020-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.delivery

import java.util.UUID

import scala.concurrent.duration._
import scala.util.Failure
import scala.util.Success

import akka.Done
import akka.actor.typed.ActorRef
import scala.annotation.nowarn

@nowarn("msg=never used")
object WorkPullingDocExample {

  //#imports
  import akka.actor.typed.scaladsl.Behaviors
  import akka.actor.typed.Behavior
  //#imports

  //#consumer
  import akka.actor.typed.delivery.ConsumerController
  import akka.actor.typed.receptionist.ServiceKey

  object ImageConverter {
    sealed trait Command
    final case class ConversionJob(resultId: UUID, fromFormat: String, toFormat: String, image: Array[Byte])
    private case class WrappedDelivery(d: ConsumerController.Delivery[ConversionJob]) extends Command

    val serviceKey = ServiceKey[ConsumerController.Command[ConversionJob]]("ImageConverter")

    def apply(): Behavior[Command] = {
      Behaviors.setup { context =>
        val deliveryAdapter =
          context.messageAdapter[ConsumerController.Delivery[ConversionJob]](WrappedDelivery(_))
        val consumerController =
          context.spawn(ConsumerController(serviceKey), "consumerController")
        consumerController ! ConsumerController.Start(deliveryAdapter)

        Behaviors.receiveMessage {
          case WrappedDelivery(delivery) =>
            val image = delivery.message.image
            val fromFormat = delivery.message.fromFormat
            val toFormat = delivery.message.toFormat
            // convert image...
            // store result with resultId key for later retrieval

            // and when completed confirm
            delivery.confirmTo ! ConsumerController.Confirmed

            Behaviors.same
        }

      }
    }

  }
  //#consumer

  //#producer
  import akka.actor.typed.delivery.WorkPullingProducerController
  import akka.actor.typed.scaladsl.ActorContext
  import akka.actor.typed.scaladsl.StashBuffer

  object ImageWorkManager {
    sealed trait Command
    final case class Convert(fromFormat: String, toFormat: String, image: Array[Byte]) extends Command
    private case class WrappedRequestNext(r: WorkPullingProducerController.RequestNext[ImageConverter.ConversionJob])
        extends Command

    final case class GetResult(resultId: UUID, replyTo: ActorRef[Option[Array[Byte]]]) extends Command

    //#producer

    //#ask
    final case class ConvertRequest(
        fromFormat: String,
        toFormat: String,
        image: Array[Byte],
        replyTo: ActorRef[ConvertResponse])
        extends Command

    sealed trait ConvertResponse
    final case class ConvertAccepted(resultId: UUID) extends ConvertResponse
    case object ConvertRejected extends ConvertResponse
    final case class ConvertTimedOut(resultId: UUID) extends ConvertResponse

    private final case class AskReply(resultId: UUID, originalReplyTo: ActorRef[ConvertResponse], timeout: Boolean)
        extends Command
    //#ask

    //#producer
    def apply(): Behavior[Command] = {
      Behaviors.setup { context =>
        val requestNextAdapter =
          context.messageAdapter[WorkPullingProducerController.RequestNext[ImageConverter.ConversionJob]](
            WrappedRequestNext(_))
        val producerController = context.spawn(
          WorkPullingProducerController(
            producerId = "workManager",
            workerServiceKey = ImageConverter.serviceKey,
            durableQueueBehavior = None),
          "producerController")
        //#producer
        //#durable-queue
        import akka.persistence.typed.delivery.EventSourcedProducerQueue
        import akka.persistence.typed.PersistenceId

        val durableQueue =
          EventSourcedProducerQueue[ImageConverter.ConversionJob](PersistenceId.ofUniqueId("ImageWorkManager"))
        val durableProducerController = context.spawn(
          WorkPullingProducerController(
            producerId = "workManager",
            workerServiceKey = ImageConverter.serviceKey,
            durableQueueBehavior = Some(durableQueue)),
          "producerController")
        //#durable-queue
        //#producer
        producerController ! WorkPullingProducerController.Start(requestNextAdapter)

        Behaviors.withStash(1000) { stashBuffer =>
          new ImageWorkManager(context, stashBuffer).waitForNext()
        }
      }
    }

  }

  //#ask

  final class ImageWorkManager(
      context: ActorContext[ImageWorkManager.Command],
      stashBuffer: StashBuffer[ImageWorkManager.Command]) {
    //#ask
    import ImageWorkManager._

    private def waitForNext(): Behavior[Command] = {
      Behaviors.receiveMessagePartial {
        case WrappedRequestNext(next) =>
          stashBuffer.unstashAll(active(next))
        case c: Convert =>
          if (stashBuffer.isFull) {
            context.log.warn("Too many Convert requests.")
            Behaviors.same
          } else {
            stashBuffer.stash(c)
            Behaviors.same
          }
        case GetResult(resultId, replyTo) =>
          // TODO retrieve the stored result and reply
          Behaviors.same
      }
    }

    private def active(
        next: WorkPullingProducerController.RequestNext[ImageConverter.ConversionJob]): Behavior[Command] = {
      Behaviors.receiveMessagePartial {
        case Convert(from, to, image) =>
          val resultId = UUID.randomUUID()
          next.sendNextTo ! ImageConverter.ConversionJob(resultId, from, to, image)
          waitForNext()
        case GetResult(resultId, replyTo) =>
          // TODO retrieve the stored result and reply
          Behaviors.same
        case _: WrappedRequestNext =>
          throw new IllegalStateException("Unexpected RequestNext")
      }
    }
    //#producer
    object askScope {
      //#ask

      import WorkPullingProducerController.MessageWithConfirmation
      import akka.util.Timeout

      implicit val askTimeout: Timeout = 5.seconds

      private def waitForNext(): Behavior[Command] = {
        Behaviors.receiveMessagePartial {
          case WrappedRequestNext(next) =>
            stashBuffer.unstashAll(active(next))
          case c: ConvertRequest =>
            if (stashBuffer.isFull) {
              c.replyTo ! ConvertRejected
              Behaviors.same
            } else {
              stashBuffer.stash(c)
              Behaviors.same
            }
          case AskReply(resultId, originalReplyTo, timeout) =>
            val response = if (timeout) ConvertTimedOut(resultId) else ConvertAccepted(resultId)
            originalReplyTo ! response
            Behaviors.same
          case GetResult(resultId, replyTo) =>
            // TODO retrieve the stored result and reply
            Behaviors.same
        }
      }

      private def active(
          next: WorkPullingProducerController.RequestNext[ImageConverter.ConversionJob]): Behavior[Command] = {
        Behaviors.receiveMessagePartial {
          case ConvertRequest(from, to, image, originalReplyTo) =>
            val resultId = UUID.randomUUID()
            context.ask[MessageWithConfirmation[ImageConverter.ConversionJob], Done](
              next.askNextTo,
              askReplyTo =>
                MessageWithConfirmation(ImageConverter.ConversionJob(resultId, from, to, image), askReplyTo)) {
              case Success(done) => AskReply(resultId, originalReplyTo, timeout = false)
              case Failure(_)    => AskReply(resultId, originalReplyTo, timeout = true)
            }
            waitForNext()
          case AskReply(resultId, originalReplyTo, timeout) =>
            val response = if (timeout) ConvertTimedOut(resultId) else ConvertAccepted(resultId)
            originalReplyTo ! response
            Behaviors.same
          case GetResult(resultId, replyTo) =>
            // TODO retrieve the stored result and reply
            Behaviors.same
          case _: WrappedRequestNext =>
            throw new IllegalStateException("Unexpected RequestNext")
        }
      }

      //#ask
    }
    //#producer
    //#ask
  }
  //#ask
  //#producer

}
